File: day5/non_constructible_change.java
/* Resumed from 01/01/2023
Time Complexity: O(nlogn) -> As it involves sorting the array too.
Space Complexity: O(1) -> As same coins array is sorted. */

import java.util.*;

class Program {

  public int nonConstructibleChange(int[] coins) {
    Arrays.sort(coins);
    int coinChange = 0;
    for (int i=0;i < coins.length; i++) {
      if(coins[i] > coinChange + 1) {
        break;
      }
      coinChange += coins[i];
    }
    return coinChange + 1;
  }
}

File: day5/non_constructible_change.py
# Resumed from 01/01/2023
# Time Complexity: O(nlogn) -> As it involves sorting the array too.
# Space Complexity: O(1) -> As same coins array is sorted.
def nonConstructibleChange(coins):
    coins.sort()
    coinChange = 0
    for coin in coins:
        if coin > coinChange + 1:
            return coinChange + 1
        coinChange += coin
    return coinChange + 1
